launcher.fnl

```diff
@@ -0,0 +1,172 @@
+  for production. It is used for docstrings and enabled by default in the REPL.
+
+  When not given a command, runs the file given as the first argument.
+  When given neither command nor file, launches a repl.
+
+  If ~/.fennelrc exists, loads it before launching a repl.")
+
+(local options [])
+
+(fn dosafely [f ...]
+  (let [args [...]
+        (ok val) (xpcall #(f (unpack args)) fennel.traceback)]
+    (when (not ok)
+      (io.stderr:write (.. val "\n"))
+      (os.exit 1))
+    val))
+
+(fn allow-globals [global-names]
+  (set options.allowedGlobals [])
+  (each [g (global-names:gmatch "([^,]+),?")]
+    (table.insert options.allowedGlobals g)))
+
+(for [i (# arg) 1 -1]
+  (match (. arg i)
+    "--no-searcher" (do (set options.no_searcher true)
+                        (table.remove arg i))
+    "--indent" (do (set options.indent (table.remove arg (+ i 1)))
+                   (when (= options.indent "false")
+                     (set options.indent false))
+                   (table.remove arg i))
+    "--add-package-path" (let [entry (table.remove arg (+ i 1))]
+                           (set package.path (.. entry ";" package.path))
+                           (table.remove arg i))
+    "--add-fennel-path" (let [entry (table.remove arg (+ i 1))]
+                          (set fennel.path (.. entry ";" fennel.path))
+                          (table.remove arg i))
+    "--load" (let [file (table.remove arg (+ i 1))]
+               (dosafely fennel.dofile file options [])
+               (table.remove arg i))
+    "--no-fennelrc" (do (set options.fennelrc false)
+                        (table.remove arg i))
+    "--correlate" (do (set options.correlate true)
+                      (table.remove arg i))
+    "--check-unused-locals" (do (set options.checkUnusedLocals true)
+                                (table.remove arg i))
+    "--globals" (do (allow-globals (table.remove arg (+ i 1)))
+                    (each [global-name (pairs _G)]
+                      (table.insert options.allowedGlobals global-name))
+                    (table.remove arg i))
+    "--globals-only" (do (allow-globals (table.remove arg (+ i 1)))
+                         (table.remove arg i))
+    "--require-as-include" (do (set options.requireAsInclude true)
+                               (table.remove arg i))
+    "--metadata" (do (set options.useMetadata true)
+                     (table.remove arg i))
+    "--no-metadata" (do (set options.useMetadata false)
+                        (table.remove arg i))))
+
+(when (not options.no_searcher)
+  (let [opts []]
+    (each [k v (pairs options)]
+      (tset opts k v))
+    (table.insert (or package.loaders package.searchers)
+                  (fennel.make_searcher opts))))
+
+(fn try-readline [ok readline]
+  (when ok
+    (when readline.set_readline_name
+      (readline.set_readline_name "fennel"))
+    (readline.set_options {:keeplines 1000 :histfile ""})
+    (fn opts.readChunk [parser-state]
```

Comment:
Currently fails due to `opts` not being defined:

```suggestion
    (fn options.readChunk [parser-state]
```

Reply:
How ... does this even compile???

Reply:
Yeah, this is a bug in the compiler; it fails to catch unknown globals in this case.